test(data-processor): add tests for cloud logs data processor

azizsonawalla opened the pull request:
Adds tests for the Cloud Logs Data Processor. No implementation yet for that processor, but the test cases were long enough that I think they should be their own PR.

azizsonawalla commented on packages/monitoring-system/data-processor/src/data-processor-factory.ts:
```diff
@@ -46,19 +50,26 @@ export class DataProcessorFactory implements Factory {
 
+  private getLogsProcessorOptions(): CloudLogsProcessorOptions {
+    return {
+      subscription: new PubSub().subscription('TODO'),
```
will add actual subscription name via config later

azizsonawalla commented on packages/monitoring-system/data-processor/src/data-processors/github-data-processor.ts:
```diff
@@ -104,33 +104,37 @@ export class GitHubProcessor extends DataProcessor {
-        }
-        return gitHubEvents;
-      });
+    return (
```
I think the linter changed this file to fix indentation -- no code changes here

chingor13 commented on packages/monitoring-system/data-processor/test/unit/data-processors/cloud-logs-data-processor.spec.ts:
```diff
@@ -0,0 +1,601 @@
+
+        describe('when no execution record exists', () => {
+          it('creates a new execution record and stores execution start logs', () => {
+            return testValidMessage(VALID_MESSAGES.execution_start, [
```
This method returns a promise - do we need to await all of these to ensure the assertion runs/completes?

azizsonawalla replied:
If you return a promise mocha automatically waits for it's completion. It's equivalent to doing `await` and then `done()`

chingor13 replied:
Gotcha, TIL.

chingor13 commented on packages/monitoring-system/data-processor/test/unit/data-processors/cloud-logs-data-processor.spec.ts:
```diff
@@ -0,0 +1,601 @@
+  expectedRecords: MockRecord[],
+  preExistingRecords?: MockRecord[]
+): Promise<void> {
+  if (preExistingRecords) {
+    preExistingRecords.forEach(record => mockFirestore.addRecord(record));
+  }
```
I would probably do this in the preamble of the test - setting up the conditions of the test rather than putting it in this test method.

chingor13 commented on packages/monitoring-system/data-processor/test/unit/data-processors/cloud-logs-data-processor.spec.ts:
```diff
@@ -0,0 +1,601 @@
+let mockFirestore: MockFirestore;
+let processor: CloudLogsProcessor;
+
+const VALID_MESSAGES: {[name: string]: {}} = loadFixture(
```
I might split each of these into their own file & constant - the aggregate of these don't seem to mean anything. Also, it might make sense to load it closer to the test it's used in to make the individual tests easier to read.

azizsonawalla replied:
Sure, I can do that

azizsonawalla commented on packages/monitoring-system/data-processor/test/unit/data-processors/cloud-logs-data-processor.spec.ts:
```diff
@@ -0,0 +1,601 @@
+  describe('collectAndProcess()', () => {
+    beforeEach(() => {
+      mockFirestore = new MockFirestore({
+        Bot_Execution: {},
+        Error: {},
+        Trigger: {},
+        Action: {},
+        Action_Type: {},
+        GitHub_Repository: {},
+        GitHub_Object: {},
```
Are you asking if the default should be empty collections or if this should be the default data in `MockFirestore` class?

chingor13 replied:
The latter, should MockFirestore handle this for you?

azizsonawalla replied:
`MockFirestore` is designed as a general purpose mock instead of a mock for the Data Processor's Firestore. So I didn't want to add Data Processor specific things into `MockFirestore` itself because I wanted to keep it as reusable as possible.

That being said, I could also create a `MockDataProcessorFirestore` that `extends MockFirestore` to add this default data, but it might just add more layers and complexity.

azizsonawalla commented on packages/monitoring-system/data-processor/test/unit/data-processors/cloud-logs-data-processor.spec.ts:
```diff
@@ -0,0 +1,601 @@
+              bot_id: 'merge_on_green',
+              start_time: 1595536893000,
+              logs_url:
+                'https://pantheon.corp.google.com/logs/query;query=resource.type%3D%22' +
```
Sure! I didn't realize it automatically redirected!

azizsonawalla commented on packages/monitoring-system/data-processor/test/unit/data-processors/cloud-logs-data-processor.spec.ts:
```diff
@@ -0,0 +1,601 @@
+          document: {
+            '4ww4alqs7ikq': {
+              execution_id: '4ww4q2vqvkl1',
+              bot_id: 'auto_label',
```
good catch!

azizsonawalla commented on packages/monitoring-system/data-processor/test/unit/data-processors/github-data-processor.spec.ts:
```diff
@@ -38,12 +38,17 @@ let mockFirestoreData1: FirestoreData;
 
 function resetMockData() {
```
I changed `loadFixture`'s `copy` param to be `false` by default instead of true so had to adjust parameters here.

azizsonawalla commented on packages/monitoring-system/data-processor/test/unit/data-processors/cloud-tasks-data-processor.spec.ts:
```diff
@@ -40,10 +40,22 @@ describe('Cloud Tasks Data Processor', () => {
   let MockFirestoreData3: FirestoreData;
 
   function resetMockData() {
```
I changed `loadFixture`'s `copy` param to be `false` by default instead of `true` so had to adjust parameters here.

azizsonawalla commented:
Changes from feedback:
* Fixture files are broken-down more and organized into separate folders
* Fixtures are loaded closer to the test itself
* Fixed some issues with expected test results
* Added some missing return statements

azizsonawalla commented:
Unfortunately there's not really any way around that while also keeping these tests 'black-box'. Even using something like `sinon` to stub firestore would require the same data merging to be implemented within the stub and that would in fact be messier code.

The alternative would be to use spies to ensure that `firestore.update()` was called instead of `firestore.set()` but that would make these tests whitebox and more fragile/prone to breaking if the implementation is changed
